EL 15 - SuccessFactors - Performance Feedback

VenessaFernandes-GSLab opened the pull request:
## Highlights
Added churros tests for resources
- achievements
- activities
- continuous-feedback
- continuous-feedback-request
- goal-templates
- goals
- permissions
- user-accounts/{id}/permissions

## Screenshot
![churros-pass-30jan-1](https://user-images.githubusercontent.com/33524101/51964053-56fd3680-248b-11e9-8072-907d89b9c35a.JPG)
![churros-pass-30jan-2](https://user-images.githubusercontent.com/33524101/51964052-5664a000-248b-11e9-9aaa-87eaf43ffe5e.JPG)


## Reference
* [soba](https://github.com/cloud-elements/soba/pull/9943)
* [sauce](https://github.com/cloud-elements/churros-sauce/pull/277)

## Closes
* [EL-15](https://cloudelements.atlassian.net/browse/EL-15)

MayurDhande-GSLab commented on src/test/elements/successfactors/achievements.js:
```diff
@@ -0,0 +1,22 @@
+const suite = require('core/suite');
+const cloud = require('core/cloud');
+const tools = require('core/tools');
+
```
remove this line

MayurDhande-GSLab commented on src/test/elements/successfactors/achievements.js:
```diff
@@ -0,0 +1,22 @@
+const achievementsUpdatePayload = tools.requirePayload(`${__dirname}/assets/achievements-update.json`);
+
+suite.forElement('humancapital', 'achievements', { payload: achievementsCreatePayload }, (test) => {
+  it('should allow CRUD for /hubs/humancapital/achievements', () => {
```
use test.should.supportcrud
https://cloud-elements.github.io/churros/module-core_suite.test.should.html#.supportCrud

MayurDhande-GSLab commented on src/test/elements/successfactors/achievements.js:
```diff
@@ -0,0 +1,22 @@
+    let achievementId;
+    return cloud.post('/hubs/humancapital/achievements', achievementsCreatePayload)
+      .then(r => {
+        achievementId = r.body.achievementId;
```
Post transformations for this achievmentId

MayurDhande-GSLab commented on src/test/elements/successfactors/activities.js:
```diff
@@ -0,0 +1,39 @@
+const suite = require('core/suite');
+const cloud = require('core/cloud');
+const tools = require('core/tools');
+
```
remove this line

MayurDhande-GSLab commented on src/test/elements/successfactors/activities.js:
```diff
@@ -0,0 +1,39 @@
+
+suite.forElement('humancapital', 'activities', { payload: activitiesCreatePayload }, (test) => {
+        let activityId, activityFeedbackId;
+  it('should allow CRUDS for /hubs/humancapital/activities', () => {
```
use should support curds

MayurDhande-GSLab commented on src/test/elements/successfactors/activities.js:
```diff
@@ -0,0 +1,39 @@
+        activityId = r.body.activityId;
+          })
+      .then(r => cloud.get(`/hubs/humancapital/activities/${activityId}`))
+      .then(r => cloud.put(`/hubs/humancapital/activities/${activityId}`,activitiesUpdatePayload))
```
use chakram.put in shouldsupportcruds for this PUT

MayurDhande-GSLab commented on src/test/elements/successfactors/activities.js:
```diff
@@ -0,0 +1,39 @@
+      .then(r => {
+        activityFeedbackId = r.body.activityFeedbackId;
+          })
+      .then(r => cloud.get(`/hubs/humancapital/activities/${activityId}/activity-feedback`))
```
write a separate test for feedback APIs.

MayurDhande-GSLab commented on src/test/elements/successfactors/activities.js:
```diff
@@ -0,0 +1,39 @@
+  test.should.supportPagination();
+  test
+    .withOptions({ qs: { where: 'activityPriority  = \'medium\'' } })
+    .withName('should support search by activityPriority')
```
add expect case for where clause

MayurDhande-GSLab commented on src/test/elements/successfactors/activities.js:
```diff
@@ -0,0 +1,39 @@
+    .should.return200OnGet();
+    test.withOptions({ qs: { select: 'activityName' } })
+    .withName('should support select values')
+    .should.return200OnGet();
```
add expect case for where clause

MayurDhande-GSLab commented on src/test/elements/successfactors/continuous-feedback-request.js:
```diff
@@ -0,0 +1,28 @@
+const cloud = require('core/cloud');
+const tools = require('core/tools');
+const expect = require('chakram').expect;
+
```
remove this line

MayurDhande-GSLab commented on src/test/elements/successfactors/continuous-feedback-request.js:
```diff
@@ -0,0 +1,28 @@
+const tools = require('core/tools');
+const expect = require('chakram').expect;
+
+const continuousFeedbackRequestCreatePayload = tools.requirePayload(`${__dirname}/assets/continuousFeedbackRequest-create.json`);
```
I see your filenames are inconsistent across the object names.
You have given filenames in camel cases and somewhere you have separated using `-`.
Please  check for the proper CE convention and update filenames,
make sure churros are working with those changes.

MayurDhande-GSLab commented on src/test/elements/successfactors/continuous-feedback-request.js:
```diff
@@ -0,0 +1,28 @@
+const continuousFeedbackRequestCreatePayload = tools.requirePayload(`${__dirname}/assets/continuousFeedbackRequest-create.json`);
+
+suite.forElement('humancapital', 'continuous-feedback-request', { payload: continuousFeedbackRequestCreatePayload }, (test) => {
+  it(`should allow CR for ${test.api}`, () => {
```
use should.supportcr

MayurDhande-GSLab commented on src/test/elements/successfactors/continuous-feedback.js:
```diff
@@ -0,0 +1,31 @@
+const cloud = require('core/cloud');
+const tools = require('core/tools');
+const expect = require('chakram').expect;
+
```
remove this line

MayurDhande-GSLab commented on src/test/elements/successfactors/continuous-feedback.js:
```diff
@@ -0,0 +1,31 @@
+const continuousFeedbackUpdatePayload = tools.requirePayload(`${__dirname}/assets/continuousFeedback-update.json`);
+
+suite.forElement('humancapital', 'continuous-feedback', { payload: continuousFeedbackCreatePayload }, (test) => {
+  //test.should.supportCrus();
```
remove comments from js files.
If CRUS test needs to be skipped, use skip:true

MayurDhande-GSLab commented on src/test/elements/successfactors/goal-template.js:
```diff
@@ -0,0 +1,17 @@
+'use strict';
+const suite = require('core/suite');
```
add a new line after use strict statement

MayurDhande-GSLab commented on src/test/elements/successfactors/goals.js:
```diff
@@ -0,0 +1,23 @@
+const suite = require('core/suite');
+const cloud = require('core/cloud');
+const tools = require('core/tools');
+
```
remove this line

MayurDhande-GSLab commented on src/test/elements/successfactors/permissions.js:
```diff
@@ -0,0 +1,19 @@
+
+const suite = require('core/suite');
+const expect = require('chakram').expect;
+
```
remove this line

MayurDhande-GSLab commented on src/test/elements/successfactors/user-accounts.js:
```diff
@@ -9,7 +9,7 @@ const userAccountsCreatePayload = tools.requirePayload(`${__dirname}/assets/user
 suite.forElement('humancapital', 'user-accounts', null, (test) => {
-  let id;
+  let id = '03b7ea14-2528-4138-891b-d19a0b62596a';
   it.skip(`should allow Create for ${test.api}`, () => {
```
add a comment explaining why this test was skipped

VenessaFernandes-GSLab replied:
This was a previously written test and was possibly skipped due to EL-1076 as given on line 17 of the same file. My changes include adding tests for sub-resource `permissions`

VenessaFernandes-GSLab commented:
- Formatted code
- Made use of supportcruds in achievements, activities, continuous-feedback, continuous-feedback-request
- Made asset filenames uniform as per CE convention
- Added filter validation in achievements
- Posted transformations for achievements and activities

tyler-cloud-elements commented on src/test/elements/successfactors/achievements.js:
```diff
@@ -0,0 +1,31 @@
+  test.withOptions({ qs: { select: 'achievementName' } })
+    .withName('should support select values')
+	.withValidation((r) => {		
+          expect(r.body[0]).to.contain.key('achievementName');      
```
Would be great if we could beef this up a bit to say,
```suggestion
          expect(r.body[0]).to.contain.key('achievementName').but.not.contain.key('achievementId');      
```

tyler-cloud-elements commented on src/test/elements/successfactors/activities.js:
```diff
@@ -0,0 +1,51 @@
+  test.withOptions({ qs: { select: 'activityName' } })
+    .withName('should support select values')
+	.withValidation((r) => {		
+          expect(r.body[0]).to.contain.key('activityName');      
```
Same comment as above, would be great to make sure we're only getting that key

tyler-cloud-elements commented on src/test/elements/successfactors/continuous-feedback-request.js:
```diff
@@ -0,0 +1,20 @@
+const continuousFeedbackRequestCreatePayload = tools.requirePayload(`${__dirname}/assets/continuousFeedbackRequest-create.json`);
+
+suite.forElement('humancapital', 'continuous-feedback-request', { payload: continuousFeedbackRequestCreatePayload }, (test) => {
+  //Skipping as there is no delete API supported
```
😢

tyler-cloud-elements commented on src/test/elements/successfactors/user-accounts.js:
```diff
@@ -9,7 +9,7 @@ const userAccountsCreatePayload = tools.requirePayload(`${__dirname}/assets/user
 
 suite.forElement('humancapital', 'user-accounts', null, (test) => {
-  let id;
+  let id = '03b7ea14-2528-4138-891b-d19a0b62596a';
```
Can we move this to the sauce file, i.e. 
```suggestion
  let id = require('core/props').getForKey('successfactors', 'userAccountId');
```

VenessaFernandes-GSLab commented:
- Changed `select` parameter to `fields` parameter
- Enhanced validation for `fields`
- Moved id value used in user-accounts to sauce file
- Modified validations for changed resource `goals/{externalCode}/activities`

tyler-cloud-elements commented on src/test/elements/successfactors/user-accounts.js:
```diff
@@ -9,7 +9,8 @@ const userAccountsCreatePayload = tools.requirePayload(`${__dirname}/assets/user
 suite.forElement('humancapital', 'user-accounts', null, (test) => {
-  let id;
+  //let id = '03b7ea14-2528-4138-891b-d19a0b62596a';
+  let id = require('core/props').getForKey('successfactors5', 'userAccountId');
```
`successfactors5` seems off?

tyler-cloud-elements commented on src/test/elements/successfactors/user-accounts.js:
```diff
@@ -9,7 +9,8 @@ const userAccountsCreatePayload = tools.requirePayload(`${__dirname}/assets/user
 
 suite.forElement('humancapital', 'user-accounts', null, (test) => {
-  let id;
+  //let id = '03b7ea14-2528-4138-891b-d19a0b62596a';
```
can remove this

tyler-cloud-elements commented on src/test/elements/successfactors/user-accounts.js:
```diff
@@ -21,8 +22,7 @@ suite.forElement('humancapital', 'user-accounts', null, (test) => {
   });
-  
-  //skipping due to EL-1076
+
```
Go ahead and leave this comment 

```suggestion
//skipping due to EL-1076
```
